### PR TITLE
Disable Current Of for replicated table

### PR DESCRIPTION
**Pull request by pengzhout**
Let's list the plans using CURRENT OF:

    explain delete from t1 WHERE CURRENT OF c1;
                            QUERY PLAN
    -----------------------------------------------------------
     Delete on t1  (cost=0.00..100.01 rows=1 width=10)
       ->  Tid Scan on t1  (cost=0.00..100.01 rows=1 width=10)
             Output: ctid, gp_segment_id
             TID Cond: CURRENT OF c1
             Filter: CURRENT OF c1

    explain UPDATE r1 SET b = 3 WHERE CURRENT OF c1;
                            QUERY PLAN
    -----------------------------------------------------------
     Update on r1  (cost=0.00..100.01 rows=1 width=18)
       ->  Tid Scan on r1  (cost=0.00..100.01 rows=1 width=18)
             Output: ctid, gp_segment_id
             TID Cond: CURRENT OF c1
             Filter: CURRENT OF c1

(ctid + gp_segment_id) can identify a unique row for hash distributed
table or randomly distributed table, but for replicated table, it can
only identify a row in one replica, so in this case other replicas
are not updated or deleted properly. There is no proper way to let
CURRENT OF work fine for replicated table, so just disable it for now.

Updatable view works fine now because all the operations are designed
to executed locally, rows of each replica are never flowed to other
segments and each replica have the same data flow. For example:
     create view v_r2 as select * from r2 where c2 = 3;

     explain update v_r2 set c2 = 4 from p1 where p1.c2=v_r2.c2;
                            QUERY PLAN
    -----------------------------------------------------------
     Update on r2
       ->  Nested Loop
             ->  Seq Scan on r2
                   Filter: (c2 = 3)
             ->  Materialize
                   ->  Broadcast Motion 3:3
                         ->  Seq Scan on p1
                               Filter: (c2 = 3)
As the plan shows, p1 is broadcasted, so all replicas of replicated
table operate locally.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

### CODE REVIEW

File: src/backend/parser/parse_relation.c [hunk 1]
```diff
@@ -1980,6 +1980,17 @@ isSimplyUpdatableRelation(Oid relid, bool noerror)
 			return_value = false;
 			break;
 		}
+
+		if (GpPolicyIsReplicated(rel->rd_cdbpolicy))
```
Comment (hlinnaka) [inline src/backend/parser/parse_relation.c]: A brief comment would be good, to explain why replicated tables are not supported.

File: src/backend/parser/parse_relation.c [hunk 2]
```diff
@@ -1980,6 +1980,17 @@ isSimplyUpdatableRelation(Oid relid, bool noerror)
...
+		{
+			if (!noerror)
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("\"%s\" is not simply updatable",
+								RelationGetRelationName(rel))));
+			return_value = false;
+			break;
+		}
 	} while (0);
```
Comment (paul-guo-) [inline src/backend/parser/parse_relation.c]: do while(0) seems to be ugly here. Maybe we should remove them along with break and rewrite with a simpler code logic like
```
if (rel->rd_rel->relkind == RELKIND_VIEW || rel->rd_rel->relkind != RELKIND_RELATION ....) 
```
given the error handling code is same.

This function is gpdb specific so there will be no concern about pg merge also.

File: src/backend/parser/parse_relation.c (hunk 1 above)
Comment (pengzhout) [reply src/backend/parser/parse_relation.c]: added.

File: src/backend/parser/parse_relation.c (hunk 2 above)
Comment (pengzhout) [reply src/backend/parser/parse_relation.c]: it seems to be introduced by commit 64581884618 for a purpose.

**Comment by pengzhout** [comment]
merged, thanks.

**Pull request merged**